spring-integration-mqtt/src/main/java/org/springframework/integration/mqtt/outbound/AbstractMqttMessageHandler.java

```diff
@@ -109,8 +198,49 @@ protected void incrementClientInstance() {
 	@Override
 	protected void onInit() throws Exception {
 		super.onInit();
+		if (this.topicProcessor instanceof BeanFactoryAware && getBeanFactory() != null) {
+			((BeanFactoryAware) this.topicProcessor).setBeanFactory(getBeanFactory());
+		}
+		if (this.qosProcessor instanceof BeanFactoryAware && getBeanFactory() != null) {
+			((BeanFactoryAware) this.qosProcessor).setBeanFactory(getBeanFactory());
+		}
+		if (this.retainedProcessor instanceof BeanFactoryAware && getBeanFactory() != null) {
+			((BeanFactoryAware) this.retainedProcessor).setBeanFactory(getBeanFactory());
+		}
 		if (this.converter == null) {
-			this.converter = new DefaultPahoMessageConverter(this.defaultQos, this.defaultRetained);
+			DefaultPahoMessageConverter mConverter = new DefaultPahoMessageConverter(
+					m -> {
+						if (this.qosProcessor == null) {
+							return MqttMessageConverter.defaultQosFunction(this.defaultQos).apply(m);
+						}
+						else {
+							Integer result = this.qosProcessor.processMessage(m);
+							if (result == null) {
+								return this.defaultQos;
+							}
+							else {
+								return result;
+							}
+						}
+					},
```

Comment:
I try to find the proper words to say that this design doesn't look good, but still failing.
Maybe the `this.qosProcessor` can be just of that `defaultQosFunction`:
```
private MessageProcessor<Integer> qosProcessor = MqttMessageConverter.defaultQosFunction(this.defaultQos)::apply;
```
?
Although from there it looks like that `defaultQosFunction()` doesn't need to have `this.defaultQos` fallback option, because we have it here in the current lambda. But looking further we don't need that  `defaultQosFunction()` , because we can simply have `qosProcessor` as a function of ` message.getHeaders().get(MqttHeaders.QOS)` and fallback to `this.defaultQos` here again.

I don't say that the logic is wrong, but I just see the design a bit differently...

Huh! That is my five fingers :smile: